### PR TITLE
i128 and u128 support

### PR DESCRIPTION
**Pull request by est31**
Brings i128 and u128 support to nightly rust, behind a feature flag. The goal of this PR is to do the bulk of the work for 128 bit integer support. Smaller but just as tricky features needed for stabilisation like 128 bit enum discriminants are left for future PRs.

Rebased version of  #37900, which in turn was a rebase + improvement of #35954 . Sadly I couldn't reopen #37900 due to github. There goes my premium position in the homu queue...

[plugin-breaking-change]

cc #35118 (tracking issue)

### CODE REVIEW

**Comment by est31** [comment]
I've removed the workaround for #37686 as the bootstrap compiler got updated.

r? @eddyb

**Comment by retep998** [comment]
>Sadly I couldn't reopen #37900 due to github.

If you push to a PR branch while the PR is closed, github won't let you reopen it. You have to reopen it first, then push the new version.

**Comment by est31** [comment]
@retep998 yes, found that out after I've already pushed it. :/

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -357,33 +373,37 @@ impl Integer {
...
-            -0x0000_0080...0x0000_007f => I8,
-            -0x0000_8000...0x0000_7fff => I16,
-            -0x8000_0000...0x7fff_ffff => I32,
-            _ => I64
+            -0x0000_0000_0000_0001...0x0000_0000_0000_0000 => I1,
+            -0x0000_0000_0000_0080...0x0000_0000_0000_007f => I8,
+            -0x0000_0000_0000_8000...0x0000_0000_0000_7fff => I16,
+            -0x0000_0000_8000_0000...0x0000_0000_7fff_ffff => I32,
+            -0x8000_0000_0000_0000...0x7fff_ffff_ffff_ffff => I64,
+            _ => I128
```
Comment (samueltardieu) [inline src/librustc/ty/layout.rs]: Match argument is an `i64` there, `I128` will never be returned.

File: src/librustc/ty/layout.rs [hunk 2]
```diff
@@ -357,33 +373,37 @@ impl Integer {
...
-            0...0x0000_00ff => I8,
-            0...0x0000_ffff => I16,
-            0...0xffff_ffff => I32,
-            _ => I64
+            0...0x0000_0000_0000_0001 => I1,
+            0...0x0000_0000_0000_00ff => I8,
+            0...0x0000_0000_0000_ffff => I16,
+            0...0x0000_0000_ffff_ffff => I32,
+            0...0xffff_ffff_ffff_ffff => I64,
+            _ => I128,
```
Comment (samueltardieu) [inline src/librustc/ty/layout.rs]: Match argument is an `u64` there, `I128` will never be returned.

**Review by samueltardieu — COMMENTED** [review]
I read the diff out of curiosity and was surprised by this.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (sinkuu) [inline src/librustc/ty/layout.rs]: See https://github.com/rust-lang/rust/pull/37900#discussion_r90778972

**Comment by alexcrichton** [comment]
@bors: r-

Looking at some [travis logs](https://travis-ci.org/rust-lang/rust/builds/185524801) it looks like this [doesn't compile on 32-bit](https://travis-ci.org/rust-lang/rust/jobs/185524806) maybe? I think the error is:

```
error[E0053]: method `to_ret` has an incompatible type for trait
   --> /checkout/src/libcompiler_builtins/lib.rs:505:28
    |
426 |         fn to_ret(self) -> Self::Ret;
    |                            --------- type in trait
...
505 |         fn to_ret(self) -> u128ret {
    |                            ^^^^^^^ expected i128, found u128
error: aborting due to previous error
```

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (DirkyJerky) [inline src/librustc/ty/layout.rs]: Would it be better to actually add the "proper" case, but have the *128 case return a *64 instead, with a comment linking to the relevant issue?

**Comment by est31** [comment]
Rebased. r? @eddyb

**Comment by est31** [comment]
Ok, this seems like another windows ABI issue with the intrinsics, this time on 32 bit. I'm trying to reproduce the issue right now.

Fixing the bug is a matter of finding out what LLVM expects as ABI in this case and replicating it in Rust.

**Comment by est31** [comment]
@nagisa can you have a look at https://gist.github.com/est31/e2cc6731df348b0fb8660a684519956a .

I've compiled an i386 binary to do multiplication of u128.

**Comment by est31** [comment]
Rebased, but not asking for r? yet because of the windows issue still being present (see above).

**Comment by est31** [comment]
Sorry, didn't test x86_64. There were some build issues with it. Now it should be fixed. r? @eddyb

**Comment by est31** [comment]
I'll pushing some WIP code shortly. Its still failing to do `python x.py test .\src\libserialize -j 1` with:

```
test leb128::test_signed_leb128 ... error: test failed
```

Note, its actually printing an error message now.

Sadly, I'm going home now for christmas vacations, and this means I don't have access to a windows VM anymore.

**Comment by nagisa** [comment]
@eddyb this is ready for another go at bots.

**Comment by eddyb** [comment]
This adds a new unstable ABI, `"unadjusted"`, to make `compiler_builtins` work on win64.
Are we okay with that? We might eventually not need it given LLVM changes. Also cc @rust-lang/lang.

@rfcbot merge

**Comment by nagisa** [comment]
For discussion on unadjusted-ABI, the last commit is the only one relevant here. The ABI added is intended to be perma-unstable as many notes all over the related parts of code allude to.

I hope it will be possible to eventually make LLVM expect intrinsics to have proper C/win64 calling convention and in turn that this unadjusted ABI ends up eventually removed.

**Comment by eddyb** [comment]
@dikaiosune Am I doing this wrong? I haven't yet been able to get @rfcbot to do anything.

**Comment by eddyb** [comment]
@rfcbot fcp merge

**Comment by est31** [comment]
This PR implements RFC https://github.com/rust-lang/rfcs/pull/1504 which has been approved of separately. As @nagisa said, only the last commit is relevant for the actual FCP discussion about "unadjusted". The problem the new ABI is used to solve are the wrong expectations of LLVM of what the ABI on windows should look like. An [issue has been filed](https://llvm.org/bugs/show_bug.cgi?id=31362
) about this (maybe a second issue report is needed for the intrinsics, not sure).

To better talk to LLVM, the "unadjusted" ABI allows for a (comparatively) simple workaround. @nagisa had tried to implement naked function wrappers + asm that try to mimic LLVM's expectations, but this would have been a very prone to break solution (consider what happens if something in LLVM changes, and suddenly integer literals don't parse anymore on windows, people won't want to have to track this down to intrinsics ABIs).

The intention is to eventually remove (or deprecate + remove) the ABI once those LLVM issues are fixed.

**Comment by pnkfelix** [comment]
@rfcbot reviewed

**Comment by est31** [comment]
rebased.

**Comment by est31** [comment]
rebased.

**Comment by alexcrichton** [comment]
@bors: retry

* network failure

**Comment by nikomatsakis** [comment]
failure seems legitimate:

```
Documenting stage2 standalone (i686-pc-windows-gnu)
Copying stage2 std from stage1 (i686-pc-windows-gnu -> i686-pc-windows-gnu / i686-pc-windows-gnu)
Copying stage2 test from stage1 (i686-pc-windows-gnu -> i686-pc-windows-gnu / i686-pc-windows-gnu)
Copying stage2 rustc from stage1 (i686-pc-windows-gnu -> i686-pc-windows-gnu / i686-pc-windows-gnu)
Building stage2 tool error_index_generator (i686-pc-windows-gnu)
   Compiling error_index_generator v0.0.0 (file:///C:/projects/rust/src/tools/error_index_generator)
error: internal compiler error: src\librustc_trans\mir\operand.rs:82: impossible case reached
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'Box<Any>', src\librustc_errors\lib.rs:423
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: Could not compile `error_index_generator`.
To learn more, run the command again with --verbose.
```

**Comment by est31** [comment]
Okay, results on the issue so far: It is reproducible on linux as well, so not a windows ABI related issue (phew).

About where it appeared: it is not reproducible when I base the branch on commit 75f5981 , but is reproducible when I base it on 75f5981 , so seems that something in the recent rollup #38697 aided in the creation of this build failure. Quite a big rollup, quite many changes...

**Comment by est31** [comment]
Now some further progress. It isn't reproducible if I base the branch on 1c2a6f9, while it is when the branch is based on e484197. So the "bad" commit is somewhere in [this list of changes](http://paste.ubuntu.com/23714087/). Can't directly narrow down the commit range further sadly (the stuff in between likely won't build...).

However, I tried to do `full-bootstrap = true`, basing on commit e484197, and suddenly the compile worked. So the "compile rustc twice not thrice" change done in commit commit 7046fea5be3911 (and with a bugfix in commit e484197482698ee0bb) has caused this regression. @alexcrichton any ideas?

**Comment by alexcrichton** [comment]
@est31 can you reproduce the failure before those bootstrap changes with:

```
./x.py build src/tools/error_index_generator --stage 1
```

It sounds like this may be a staging problem, although I'm not quite sure precisely what. Those commits rearchitected how we performed a bootstrap, not compiling the last stage by default. That shouldn't have an effect on the compiler, but it sorta means that we use the stage1 compiler more often in a sense.

I haven't been following this PR too closely, but can you describe the staging strategy this PR is taking? Knowing that may help me help you to narrow it down.

**Comment by alexcrichton** [comment]
@bors: retry

* network failure

**Comment by est31** [comment]
problem fixed.

r? @eddyb

**Comment by jackpot51** [comment]
Awesome!

**Comment by est31** [comment]
wow, great!!

:tada: 

it was the 

# FIRST TRY

I assure you!!
http://devopsreactions.tumblr.com/post/151057980189/getting-your-build-through-the-pipeline

http://devopsreactions.tumblr.com/post/151057980189/getting-your-build-through-the-pipeline

**Comment by DirkyJerky** [comment]
This passed on the first try in 2017! (in utc + (-3)..9 ).
Only acceptable explanation is that 2016 truly was the year of... [what animal](https://goo.gl/photos/qjN2iPTn6nehtqidA)

**Pull request merged**